business/dashboards.go

```diff
@@ -243,26 +341,98 @@ func (in *DashboardsService) buildRuntimesList(namespace string, templatesNames
 		if dashboard == nil {
 			continue
 		}
-		runtime := dashboard.Spec.Runtime
-		ref := models.DashboardRef{
-			Template: dashboard.Name,
-			Title:    dashboard.Spec.Title,
-		}
-		found := false
-		for i := range runtimes {
-			rtObj := &runtimes[i]
-			if rtObj.Name == runtime {
-				rtObj.DashboardRefs = append(rtObj.DashboardRefs, ref)
-				found = true
-				break
+		runtimes = addDashboardToRuntimes(dashboard, runtimes)
+	}
+	return runtimes
+}
+
+func (in *DashboardsService) fetchMetricNames(namespace, app, version string) []string {
+	labels := fmt.Sprintf(`{namespace="%s",app="%s"`, namespace, app)
+	if version != "" {
+		labels += fmt.Sprintf(`,version="%s"`, version)
+	}
+	labels += "}"
+	metrics, err := in.prom.GetMetricsForLabels([]string{labels})
+	if err != nil {
+		log.Errorf("Runtimes discovery failed, cannot load metrics for labels: %s. Error was: %v", labels, err)
+	}
+	return metrics
+}
+
+func (in *DashboardsService) discoverRuntimesList(namespace, app, version string) []models.Runtime {
+	var metrics []string
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		metrics = in.fetchMetricNames(namespace, app, version)
+	}()
+
+	allDashboards, err := in.loadRawDashboardResources(namespace)
+	if err != nil {
+		log.Errorf("Runtimes discovery failed, cannot load dashboards in namespace %s. Error was: %v", namespace, err)
+		return []models.Runtime{}
+	}
+
+	wg.Wait()
+	return runDiscoveryMatcher(metrics, allDashboards)
+}
+
+func runDiscoveryMatcher(metrics []string, allDashboards map[string]v1alpha1.MonitoringDashboard) []models.Runtime {
+	// In all dashboards, finds the ones that match the metrics set
+	// We must exclude from the results included dashboards when both the including and the included dashboards are matching
+	runtimesMap := make(map[string]*v1alpha1.MonitoringDashboard)
+	for _, d := range allDashboards {
+		dashboard := d // sticky reference
+		matchReference := strings.TrimSpace(dashboard.Spec.DiscoverOn)
+		if matchReference != "" {
+			for _, metric := range metrics {
+				if matchReference == strings.TrimSpace(metric) {
+					if _, exists := runtimesMap[dashboard.Name]; !exists {
```

Comment:
I wonder what may happens if there is some metric shared between dashboards.
I.e. same metric in nodejs or vertx.
What dashboard would find ?

Reply:
since we don't break the `for _, d := range allDashboards` loop, all matching dashboards are going to appear if they share the same "discovery" metric. This is as I would expect...